File: go/graphql/mappers/jira_projects_mapper.go
package mappers

import (
	"errors"
	"strings"

	"atlassian-graphql/graphql/canonical"
	"atlassian-graphql/graphql/gen"
)

func ProjectWithOpsgenieTeams(cloudID string, project gen.JiraProjectNode, teams []gen.OpsgenieTeamNode) (canonical.CanonicalProjectWithOpsgenieTeams, error) {
	cloud := strings.TrimSpace(cloudID)
	if cloud == "" {
		return canonical.CanonicalProjectWithOpsgenieTeams{}, errors.New("cloudID is required")
	}

	key := strings.TrimSpace(project.Key)
	if key == "" {
		return canonical.CanonicalProjectWithOpsgenieTeams{}, errors.New("project.key is required")
	}
	name := strings.TrimSpace(project.Name)
	if name == "" {
		return canonical.CanonicalProjectWithOpsgenieTeams{}, errors.New("project.name is required")
	}

	seen := map[string]struct{}{}
	outTeams := make([]canonical.OpsgenieTeamRef, 0, len(teams))
	for _, t := range teams {
		id := strings.TrimSpace(t.ID)
		if id == "" {
			return canonical.CanonicalProjectWithOpsgenieTeams{}, errors.New("opsgenie_team.id is required")
		}
		if _, ok := seen[id]; ok {
			continue
		}
		seen[id] = struct{}{}

		teamName := strings.TrimSpace(t.Name)
		if teamName == "" {
			return canonical.CanonicalProjectWithOpsgenieTeams{}, errors.New("opsgenie_team.name is required")
		}
		outTeams = append(outTeams, canonical.OpsgenieTeamRef{ID: id, Name: teamName})
	}

	return canonical.CanonicalProjectWithOpsgenieTeams{
		Project: canonical.JiraProject{
			CloudID: cloud,
			Key:     key,
			Name:    name,
		},
		OpsgenieTeams: outTeams,
	}, nil
}

File: python/atlassian_graphql/mappers/jira_projects_mapper.py
from __future__ import annotations

from typing import Iterable, List

from atlassian_graphql.canonical_models import (
    CanonicalProjectWithOpsgenieTeams,
    JiraProject,
    OpsgenieTeamRef,
)
from atlassian_graphql.gen.jira_projects_api import JiraProjectNode, OpsgenieTeamNode


def map_project_with_opsgenie_teams(
    *,
    cloud_id: str,
    project: JiraProjectNode,
    opsgenie_teams: Iterable[OpsgenieTeamNode],
) -> CanonicalProjectWithOpsgenieTeams:
    cloud_id_clean = (cloud_id or "").strip()
    if not cloud_id_clean:
        raise ValueError("cloud_id is required")

    project_key = (project.key or "").strip()
    if not project_key:
        raise ValueError("project.key is required")

    project_name = (project.name or "").strip()
    if not project_name:
        raise ValueError("project.name is required")

    seen: set[str] = set()
    teams: List[OpsgenieTeamRef] = []
    for t in opsgenie_teams:
        team_id = (t.id or "").strip()
        if not team_id:
            raise ValueError("opsgenie_team.id is required")
        if team_id in seen:
            continue
        seen.add(team_id)

        team_name = (t.name or "").strip()
        if not team_name:
            raise ValueError("opsgenie_team.name is required")
        teams.append(OpsgenieTeamRef(id=team_id, name=team_name))

    return CanonicalProjectWithOpsgenieTeams(
        project=JiraProject(cloud_id=cloud_id_clean, key=project_key, name=project_name),
        opsgenie_teams=teams,
    )

